### PR TITLE
[inductor] Disable cudagraphs if index_put_ fallback is encountered

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105439

**TL;DR**: if lowerings.py encounters aten.index_put, it will set V.graph.cudagraphs_okay = False, which will disable cudagraphs. index_put needs to be disabled because it crashes cuda graphs.

index_put_ fallbacks fail with cuda graphs when `accumulate=True` - likely for the same reason that it fails with deterministic_algorithms_enabled:
https://github.com/pytorch/pytorch/blob/fcb7d4b35821d97142f1c0d8843dae0e98f8e965/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L730

A first attempt was just to expand the scenarios where `index_put_` is one of the disallowed kernels in utils.py: https://github.com/pytorch/pytorch/blob/2fa7d11b64e55b4935dbf60b7e5810cec992bf67/torch/_inductor/utils.py#L436-L438

However this disables cuda graphs in too many scenarios, because index_put doesn't cause issues if it gets fused, it only causes issues if the aten kernel gets called. So in the updated version of this PR, we check for fallbacks in lowerings.py and disable cudagraphs only if a fallback is encountered there.

Example of failure outside of PT2:

```python
import torch

def fn(x, y, z):
    x = torch.zeros_like(x)
    return x.index_put_([y], z, True)
    # return x + 1

x = torch.zeros((512, 512), dtype=torch.bool, device='cuda')
y = torch.arange(512, dtype=torch.int64, device='cuda')
z = torch.ones((512, 512), dtype=torch.bool, device='cuda')

s = torch.cuda.Stream()
s.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(s):
    for i in range(3):
        fn(x, y, z)
torch.cuda.current_stream().wait_stream(s)

g = torch.cuda.CUDAGraph()
with torch.cuda.graph(g):
    fn(x, y, z)
```

fails with
```
Traceback (most recent call last):
  File "/data/users/dberard/scripts/graphed_index_put.py", line 24, in <module>
    fn(x, y, z)
  File "/data/users/dberard/scripts/graphed_index_put.py", line 8, in fn
    return x.index_put_([y], z, True)
RuntimeError: CUDA error: operation not permitted when stream is capturing
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/users/dberard/scripts/graphed_index_put.py", line 24, in <module>
    fn(x, y, z)
  File "/data/users/dberard/pytorch/torch/cuda/graphs.py", line 173, in __exit__
    self.cuda_graph.capture_end()
  File "/data/users/dberard/pytorch/torch/cuda/graphs.py", line 79, in capture_end
    super().capture_end()
RuntimeError: CUDA error: operation failed due to a previous error during capture
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D47538548](https://our.internmc.facebook.com/intern/diff/D47538548)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47538548).

**Review by eellison — APPROVED** [review]
nice!

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47538548).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47538548).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47538548).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47538548).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47538548).

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -209,6 +209,8 @@ def __init__(
         ] = (
             []
         )  # This is the linemap used by the profiler to mark custom compiled kernels getting run
+        # Used if lowering encounters cases where cudagraphs are not supported
+        self.cudagraphs_okay = True
```
Comment (eellison) [inline torch/_inductor/graph.py:213]: nit: can we rename this to `disable_cudagraphs` ? this bool is not the the only condition to determine if we can cudagraph

**Review by eellison — APPROVED** [review]
nice !

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47538548).

**Comment by davidberard98** [comment]
it doesn't have internal changes - will detach the internal diff and land GH1
